libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,995 @@
+	dpos.path = extractPathFromFileURI(_json["textDocument"]["uri"].asString()).value();
+
+	if (boost::algorithm::starts_with(dpos.path, m_basePath.generic_string()))
+		dpos.path = dpos.path.substr(m_basePath.generic_string().size());
+
+	dpos.position.line = _json["position"]["line"].asInt();
+	dpos.position.column = _json["position"]["character"].asInt();
+
+	return dpos;
+}
+
+void LanguageServer::changeConfiguration(Json::Value const& _settings)
+{
+	if (_settings["evm"].isString())
+		if (auto const evmVersionOpt = EVMVersion::fromString(_settings["evm"].asString()); evmVersionOpt.has_value())
+			m_evmVersion = evmVersionOpt.value();
+
+	if (_settings["revertStrings"].isString())
+		m_revertStrings = revertStringsFromString(_settings["revertStrings"].asString()).value_or(RevertStrings::Default);
+
+	if (_settings["remapping"].isArray())
+	{
+		for (auto const& element: _settings["remapping"])
+		{
+			if (element.isString())
+			{
+				if (auto remappingOpt = ImportRemapper::parseRemapping(element.asString()); remappingOpt.has_value())
+					m_remappings.emplace_back(move(remappingOpt.value()));
+				else
+					trace("Failed to parse remapping: '"s + element.asString() + "'");
+			}
+		}
+	}
+
+	static string_view constexpr strModelCheckerContracts = "model-checker-contracts";
+	static string_view constexpr strModelCheckerEngine = "model-checker-engine";
+	static string_view constexpr strModelCheckerTargets = "model-checker-targets";
+	static string_view constexpr strModelCheckerTimeout = "model-checker-timeout";
+
+	if (_settings[strModelCheckerContracts.data()].isString())
+	{
+		string contractsStr = _settings[strModelCheckerContracts.data()].as<string>();
+		optional<ModelCheckerContracts> contracts = ModelCheckerContracts::fromString(contractsStr);
+		if (contracts)
+			m_modelCheckerSettings.contracts = move(*contracts);
+		else
+			log(fmt::format("Invalid option for {}: {}", strModelCheckerContracts, contractsStr));
+	}
+
+	if (_settings[strModelCheckerEngine.data()].isString())
+	{
+		string engineStr = _settings[strModelCheckerEngine.data()].as<string>();
+		optional<ModelCheckerEngine> engine = ModelCheckerEngine::fromString(engineStr);
+		if (engine)
+			m_modelCheckerSettings.engine = *engine;
+		else
+			log(fmt::format("Invalid for {}: {}", strModelCheckerEngine, engineStr));
+	}
+
+	if (_settings[strModelCheckerTargets.data()].isString())
+	{
+		string targetsStr = _settings[strModelCheckerTargets.data()].as<string>();
+		optional<ModelCheckerTargets> targets = ModelCheckerTargets::fromString(targetsStr);
+		if (targets)
+			m_modelCheckerSettings.targets = *targets;
+		else
+			log(fmt::format("Invalid for {}: {}", strModelCheckerTargets, targetsStr));
+	}
+
+	if (_settings[strModelCheckerTimeout.data()].isUInt())
+		m_modelCheckerSettings.timeout = _settings[strModelCheckerTargets.data()].asUInt();
+}
+
+void LanguageServer::documentContentUpdated(string const& _path, LineColumnRange _range, string const& _replacementText)
```

Comment:
I'm still not so happy about the file reader owning a copy of the source codes. The reader should just be a thin wrapper around the read callback, and the compiler stack should have a single copy of the source codes, but I'm fine keeping it like that if it is not a big performance problem.